### ISSUE TITLE
Conditionals: no way to check if boolean field exists

### ISSUE THREAD

**Issue by bradvido** [elastic/logstash#1867]
In the Logstash config language, I can't determine any way to check if a boolean field exists.

e.g. Using this conditional
`if[myfield] { ...exists... } else { ...doesn't exist... }`
The results from testing this conditional statement are:

```
[myfield] does not exist   --> false
[myfield] exists, is true  --> true
[myfield] exists, is false --> false //expected true because the field exists
```

I've asked the question here as well to reach out and see if someone else knows a way to check for boolean existence.
http://stackoverflow.com/questions/26287082/logstash-config-check-if-boolean-field-exists

**Comment by javifr** [comment]
+1 for @timbunce

**Comment by webwurst** [comment]
+1

**Comment by naisanza** [comment]
I'm running into this where I'm looking to use an if statement to see if the field contains data. If true, then: 

```
if [ip] !== "(?:%{null})|-|" {
    mutate {
    add_field => { "ip" => "%{mxIP}}
    }
}
```

Also, does the current implementation of the if statement allow the use of regex and boolean within a field value? Like such:

` if [ip] == "((?:%{null})|-|)" {`

**Comment by alfredocambera** [comment]
This feature would be very useful

**Comment by thalesfsp** [comment]
+1

**Comment by groeneveld91** [comment]
+1

**Comment by missnebun** [comment]
+1

**Comment by omercnet** [comment]
+1

**Comment by sahlex** [comment]
+1

**Comment by GoodMirek** [comment]
+1

**Comment by geoffroya** [comment]
+1

**Comment by dedemorton** [comment]
@jordansissel 
I want to add doc updates related to this issue. When I read through to the end, it sounds like most of the confusion is around the fact that testing for “does field exist?” is indistinguishable from testing for “is field true?” But it sounds like we can also improve the the docs. Can you help me get the doc right by answering the questions that bradvido raises? There are three main ones: What is the order of precedence of the operators, how does the "in" operator behave, and what are the type restrictions for specific operators (for example, do comparison operators for equality only accept numeric types?) Here's where I apologize for not testing this myself, but it would take me longer than it's worth, and it's not worth my time if you have the answers already.

**Comment by GoodMirek** [comment]
Let me comment.

The issue is that it is not possible to distinguish whether a boolean field
is false or does not exist.
If a test of a boolean field returns true, then it exists for sure and is
true.
If a test of a boolean field returns false, then either the field exists
and is false or the field does not exist. Ability to distinguish these two
states would be greatly appreciated.

Sorry if it was already clear and my comment was surplus.

Mirek Svoboda | +420 608 224 486 | Skype: xsvobo10

**Comment by jordansissel** [comment]
@timbunce I agree functions might be a cool way to solve this. I'm open to this.

**Comment by cooniur** [comment]
+1

**Comment by nescim** [comment]
+1

**Comment by zyphlar** [comment]
Nothing after 3 years?

**Comment by BarcoMasile** [comment]
Still have this problem

**Comment by darefilz** [comment]
Here you go:
You maybe get a field _foo_ of type **boolean** from some input. If the field actually exists you want to convert it to an integer. If not you want to add it to the event and initialize to value _42_.  So your code might look like this:
```
filter {
  if [foo] {
    mutate {
      convert => { "foo" => "integer" }
    }
  } else {
    mutate {
      add_field => { "foo" => 42 }
    }
  }
}
```
If _foo_ is false it's not converted but _42_ is added, which you definitely don't want.

**Comment by gentunian** [comment]
@darefilz looks that you may rethink the pipeline or control the data ingestion. The field `foo` is supposed to be an integer but sometimes you may be receiving a boolean value. However, I think you could solve this problem.

I understand that despite anything, you need `foo` field to be an integer and to be added to your pipeline. In the special case where the field does not exists or the field exists but it's false, then, you want to have the field `foo = 42`.

You may convert the field anyway despite its existence. You now know that if the field exists and it has the false value, you converted it into the `0` integer.

For instance:

```
input {
  http { port => 9090 }
}

filter {
  # boolean false is converted to 0
  mutate {
    convert => { "foo" => "integer" }
  }

  # if the field does not exists or the field exists and is false
  if ![foo] or [foo] == 0 {
    mutate {
      replace => { "foo" => 42 }
      # this is needed because logstash will set 42 as String.
      # maybe this is related to this issue: https://github.com/elastic/logstash/issues/2987
      convert => { "foo" => "integer" }
    }
  }
}

output {
  stdout {}
}
```

I know something like `if [foo] == false` would be better, but also there are a lot of approaches to accomplish it.

**Comment by gentunian** [comment]
Well, I forgot your last line:

> If foo is false it's not converted but 42 is added, which you definitely don't want.

which makes the pipeline simpler:
```
input {
  http { port => 9090 }
}

filter {
  # boolean false is converted to 0
  mutate {
    convert => { "foo" => "integer" }
  }

  # now we already know that if the field was
  # false, it will be converted to 0. So we check
  # only if the field does not exists
  if ![foo] {
    mutate {
      replace => { "foo" => 42 }
      # this is needed because logstash will set 42 as String.
      # maybe this is related to this issue: https://github.com/elastic/logstash/issues/2987
      convert => { "foo" => "integer" }
    }
  }
}

output {
  stdout {}
}
```

**Comment by nescim** [comment]
> can you post use cases where the distinction between false and the non-existing field is meaningful? I know that semantics may be in the way but there are a lot of workaround for this being a meaningless use case.

Suppose you have some records, but not all, marked with a boolean field that is used later in the pipeline and that for unmarked records one needs to invoke a process to make a decision. Maybe that process is expensive. Maybe it's not even 100% accurate.

**Comment by dodo5575** [comment]
I would really want to have the exist function!

My use case: If the boolean field "foo" exists, put it in an S3 good_data bucket, else, put it in another bad_data bucket

output {
  if [foo] {
    s3 {good_data}
  } else {
    s3 {bad_data}
  }    
}

Now, if the foo == false, this record will be put in the bad_data bucket, which is not what I want. And, I have a list of field to check. Some of them are boolean while others are not. Hard-coding to put the case where foo == false to the good_data bucket is not a clean solution.

**Comment by MakoWish** [comment]
This bug has been open for nearly 6 years?! I am running into this same issue. I have a boolean field that does not exist in all events, but I need to check if it exists. I have tried these different ways of testing

```
if [field] == true or [field] == false {
 --- do work ---
}
```
Logstash will to fail to start with `Expected one of [ \\t\\r\\n], \"#\", \"(\" at line...`

```
if [field] == 'true' or [field] == 'false' {
 --- do work ---
}
```
Never recognizes that the field exists whether or not it does. Guessing this is because the condition is checking for a string value which will not work for a boolean value. 

```
if "" in [field] {
 --- do work ---
}
```
Also does not work. Again, looking for a string value; which a boolean is not?

I cannot for the life of me figure out how to handle this.

**Comment by jessequinn** [comment]
@MakoWish i am in the same boat. I cannot get conditional testing with boolean field to work!!!

**Comment by MakoWish** [comment]
Still looking for a solution to this issue. There needs to be a way to test test for the presence of a boolean field. This issue has been open now for more than six years with no resolution. Would appreciate some movement on this.

**Comment by yodog** [comment]
+1 

bumping so the thread wont die.

**Comment by vytsci** [comment]
After finding such a thing is not resolved in more than a half of a decade, proves me that choosing logstash was a very bad decision of mine.

**Comment by gentunian** [comment]
@vytsci just a minor comment on that: Logstash is an excellent tool. If you found that the only problem you needed to solve is based on a workaround around some limitations of the tool, then yes, probably your decision of choosing logstash was just a hype rather than a decision made with proper knowledge, and with just simple scripting you will be fine.

**Comment by jkackley** [comment]
Eh, this isn't really a bug so much as an edge case syntactically - pretty trivial to work around. 
 A quick google turned up this thread https://stackoverflow.com/questions/30309096/logstash-check-if-field-exists that has several options, though the metadata one seems the most robust ).  Somewhat surprised it hasn't been marked resolved in favor of documentation clarifying how to work around this minor syntactic limitation.

**Comment by gentunian** [comment]
Note on that @jkackley (and I'm not advertising my answer) you should scroll down to the less voted answer because it's the updated one: https://stackoverflow.com/a/62616917/241750

**Comment by dedemorton** [comment]
@karenzone Sounds like a simple doc change would help users address this limitation.

**Comment by gtback** [comment]
Hi @gentunian. That's great, thanks for sharing! Would it be OK for you to sign our [CLA](https://www.elastic.co/contributor-agreement) and then we can copy your solution from StackOverflow into our documentation (with appropriate attribution, of course!)

**Comment by gentunian** [comment]
@gtback sure, no problem. Info is there to be shared and to help everybody. 👍

**Comment by dayotoro** [comment]
+1
